texar/tf/module_base.py

```diff
@@ -80,20 +50,20 @@ def __init__(self, hparams=None):
                 raise ValueError(
                     "`self._hparams` already exists. Argument `hparams` "
                     "must be set to `None` in this case.")
-        self._template = tf.make_template(self._hparams.name, self._build,
-                                          create_scope_now_=True)
-        self._unique_name = self.variable_scope.name.split("/")[-1]
-        self._trainable_variables = []
-        self._built = False
+
+    @abstractmethod
+    def call(self, inputs, *args, **kwargs):
```

Comment:
How is the previous issue about `input` addressed now?

Reply:
And how do other tools do?

Reply:
To the best of my knowledge, we have two choices for our base class: `tf.Module` and `tf.keras.layers.Layer`. As pointed out in this [TensorFlow PR](https://github.com/tensorflow/community/pull/56), "Conceptually I see Keras `Layer` as enriched Modules with additional features (e.g. `build(shapes)`/`call(inputs)` pattern, json/yaml serialization, regularizers, losses, metrics, masking etc). If these features are useful to you, or if you are using your subclass in a Keras `Model` then Keras `Layer` is a good choice.

If the additional functionality of Keras is not useful for you then you may prefer to subclass `tf.Module` which is simpler with far fewer features. If `tf.Module` is not useful you may want to subclass `Checkpointable` or just `object`".

Reply:
https://github.com/tensorflow/community/blob/master/rfcs/20190117-tf-module.md

Reply:
How is `tf.keras.layers.Layer` related to Keras `Model`? In particular, is `__call__(input, ..)` (i.e., name the 1st argument as `input`) a necessary interface in order to be compatible with Keras `Model`? If not, we may inherit `tf.keras.layers.Layer` but change this `__call__(input, ..)` interface a little bit

Reply:
`tf.keras.Model` is actually a subclass of `tf.keras.layers.Layer`, I think they have the same/similar interfaces.